### DISCUSSION TITLE
How to conditionally render rest parameters

### DISCUSSION THREAD

**Discussion by iongion** [vanjs-org/van#35] [Q&A] [answered]
For this code I just want the `rest` to render only if a condition

```typescript
function Collapse({isOpen, keepMounted} = props, ...rest: ChildDom[]) {
   return div({ class: "Collapse"}, rest);
}
```

I want the `rest` to be rendered only if `isOpen = false && keepMounted = false`, basically this would mean passing either args or undefined to perform removal from DOM when that condition occurs.

I have tried `van.bind`, but whatever form I find ends up with `Uncaught Error: You can't add a DOM Node that is already connected to document`

Tried something like this but does not appear to be working

```typescript
    div(
      { class: Classes.COLLAPSE_BODY },
      van.bind(toState(props.isOpen), (flag) => {
        console.debug({ flag });
        return flag ? div(...rest) : undefined;
      })
    )
```

This too (this one prints only once to the console)

```typescript
    van.bind(toState(props.isOpen), (flag) => {
      console.debug({ flag });
      return flag ? div({ class: Classes.COLLAPSE_BODY }, rest) : undefined;
    })
```

**Comment by Tao-VanJS** [comment]
Could you show me the full code as well as the stacktrace of the problem?

**Reply by iongion** [reply]
Sure

- https://stackblitz.com/edit/vitejs-vite-kc1gr9?file=src%2Fblueprint%2Fcomponents%2FCollapse.ts%3AL54

There are both versions there, second one is also annoying that I can't just pass the `rest` and that I have to wrap-it in a div which is not what the UI design expects.

```
van-0.11.10.debug.js:15 Uncaught Error: You can't add a DOM Node that is already connected to document
    at expect (van-0.11.10.debug.js:15:62)
    at checkChildValid (van-0.11.10.debug.js:59:3)
    at van-0.11.10.debug.js:92:48
    at Array.forEach (<anonymous>)
    at van-0.11.10.debug.js:92:31
    at Collapse.ts:57:48
    at van-0.11.10.debug.js:105:20
    at van-0.11.10.js:68:18
    at Set.forEach (<anonymous>)
    at updateDoms (van-0.11.10.js:66:55)
```

**Comment by Tao-VanJS** [comment, marked as answer]
Regarding your example https://stackblitz.com/edit/vitejs-vite-kc1gr9?file=src%2Fblueprint%2Fcomponents%2FCollapse.ts%3AL54, without knowing how `Collapse` is being called, it's not possible to say in details how did the error occur. However, by learning from the error message, I can tell the error is caused by trying to compose a DOM node that has already been connected to the document tree as the child node of something you want to build. This is an undefined behavior in VanJS, and an error with that message will be thrown when `van-{VERSION}.debug.js` is being used. You can refer to this section https://vanjs.org/tutorial#dom-nodes-already-in-the-document-tree-can-t-be-used-as-children for more details.

Here is a minimalist example that can reproduce the error:

```html
<body>
<script type="text/javascript" src="https://vanjs.org/code/van-0.11.10.nomodule.debug.js"></script>
<script>
const {div} = van.tags

const dom = div("Test")

van.add(document.body, dom)

// This line will throw an error because `dom` has already been connected to the document tree
const newDom = div(dom, "Test2")

van.add(document.body. newDom)
</script>
</body>
```
